File: String/Determine_two_string_are_anagram_or_not/SecondApproach.py
# Function to check if X and Y are anagram or not
def isAnagram(X, Y):

    # if X's length is not same as Y's, they can't be anagram 
    if len(X) != len(Y):
        return False

    # create an empty dict
    freqX = {}
    freqY = {}

    # Maintain count of each character of X in the dict
    for i in range(len(X)):
        freqX[X[i]] = freqX.get(X[i], 0) + 1

    # Maintain count of each character of Y in the dict
    for i in range(len(Y)):
        freqY[Y[i]] = freqY.get(Y[i], 0) + 1

    # return true if both dict have same content
    return freqX == freqY

if __name__ == "__main__":

    X = "silent"
    Y = "listen"

    if isAnagram(X, Y):
        print("Anagram")
    else:
        print("Not an Anagram")

File: String/Determine_two_string_are_anagram_or_not/SecondApproach.java
import java.util.HashMap;
import java.util.Map;

class SecondApproach {
	// Function to check if X and Y are anagrams or not
	public static boolean isAnagram(String X, String Y) {
		// if X's length is not same as Y's, they can't be anagram
		if (X.length() != Y.length()) {
			return false;
		}

		// create an empty map
		Map<Character, Integer> freqX = new HashMap<>();

		// maintain count of each character of X in the map
		for (char c: X.toCharArray()) {
			freqX.put(c, freqX.getOrDefault(c, 0) + 1);
		}

		// create a second map
		Map<Character, Integer> freqY = new HashMap<>();

		// maintain count of each character of Y in the map
		for (char c: Y.toCharArray()) {
			freqY.put(c, freqY.getOrDefault(c, 0) + 1);
		}

		// return true if both map has same content
		return freqX.equals(freqY);
	}

	public static void main(String[] args) {
		String X = "tommarvoloriddle";	// Tom Marvolo Riddle
		String Y = "iamlordvoldemort";	// I am Lord Voldemort

		if (isAnagram(X, Y)) {
			System.out.println("Anagram");
		} else {
			System.out.println("Not a Anagram");
		}
	}
}
